File: solutions/go/pig-latin/1/pig_latin.go
package piglatin

import (
    "strings"
)

func Sentence(sentence string) string {
	words := strings.Fields(sentence)
	for i, word := range words {
		words[i] = translateWord(word)
	}
	return strings.Join(words, " ")
}

func translateWord(word string) string {
	vowels := "aeiou"

	// Rule 1: Starts with vowel or "xr"/"yt"
	if strings.HasPrefix(word, "xr") || strings.HasPrefix(word, "yt") || strings.ContainsAny(string(word[0]), vowels) {
		return word + "ay"
	}

	// Rule 3: consonants + "qu"
	// Проверяем "qu" до обычных согласных, так как "qu" — это особый случай
	for i := 0; i < len(word)-1; i++ {
		if word[i:i+2] == "qu" {
			// Если перед qu нет гласных (или это начало слова)
			if !strings.ContainsAny(word[:i], vowels) {
				return word[i+2:] + word[:i+2] + "ay"
			}
			break // Если нашли гласную раньше qu, идем к Rule 2
		}
	}

	// Rule 4: consonants + "y"
	// Ищем "y" после одной или более согласных
	for i := 1; i < len(word); i++ {
		if word[i] == 'y' {
			if !strings.ContainsAny(word[:i], vowels) {
				return word[i:] + word[:i] + "ay"
			}
			break
		}
	}

	// Rule 2: Move consonants to the end
	for i, letter := range word {
		if strings.ContainsAny(string(letter), vowels) {
			return word[i:] + word[:i] + "ay"
		}
	}

	return word + "ay"
}

File: solutions/python/pig-latin/1/pig_latin.py
def pig_latin(word):
    vowels = "aeiou"
    if word.startswith(('xr', 'yt')) or word[0] in vowels:
        return word + 'ay'

    for i in range(len(word) - 1):
        if word[i:i + 2] == 'qu':
            if not any(char in vowels for char in word[:i]):
                return word[i + 2:] + word[:i + 2] + 'ay'
            break

    for i in range(1, len(word)):
        if word[i] == 'y':
            if not any(char in vowels for char in word[:i]):
                return word[i:] + word[:i] + 'ay'
            break

    for i, letter in enumerate(word):
        if letter in vowels:
            return word[i:] + word[:i] + 'ay'

    return word + 'ay'

def translate(text):
    words = text.split()
    pig_latin_words = [pig_latin(word) for word in words]
    return ' '.join(pig_latin_words)
    